### PR TITLE
2178: IOException Should Be Thrown When Journal Is Corrupted

### LINKED ISSUE #2178: JournalChannel should throw IOException if it cannot seek to desired position
**BUG REPORT**

***Describe the bug***

Here's a piece of code from `JournalChannel.java`
```
try {
    if (position == START_OF_FILE) {
        if (formatVersion >= V5) {
            fc.position(HEADER_SIZE);
        } else if (formatVersion >= V2) {
            fc.position(VERSION_HEADER_SIZE);
        } else {
            fc.position(0);
        }
    } else {
        fc.position(position);
    }
} catch (IOException e) {
    LOG.error("Bookie journal file can seek to position :", e); <-- Swallows IOException and moves on
}
```

***To Reproduce***

Create a corrupt journal file. Try seeking to the desired position.

***Expected behavior***

IOException should be thrown and bookie should be shutdown. Not sure if a retry would help. It's not really recoverable since journal file is corrupted. 

@reddycharan @jvrao @eolivelli Another potential problem?

### PR DESCRIPTION
**Pull request by atris**
Fixes #2178

### CODE REVIEW

**Comment by atris** [comment]
> Can you please referente the originale issue in the description of this pr.
Done
> Also we need a test that demonstrate the fix and an explanation about how it works

I wanted to check on that. Looking in BookieJournalTest, can I use writeV5Journal API to construct a corrupted file and test that the bookie closes?

**Comment by eolivelli** [comment]
Yes it can be a good place to put the test

**Comment by atris** [comment]
@eolivelli Need a bit of help with the test. I have created a test such as:
https://gist.github.com/atris/773e1074edd1d50786ff35e300495984

However, I was not able to trigger the IOException that this PR added. What am I doing wrong, please?

**Comment by sijie** [comment]
@atris I think the issue was thrown when attempting to seek to a position doesn't exist. so I guess you can truncate the file to reproduce.

**Comment by atris** [comment]
> @atris I think the issue was thrown when attempting to seek to a position doesn't exist. so I guess you can truncate the file to reproduce.

@sijie Thanks, I updated the test but FileChannel.position then moves the file pointer to the asked location. What am I missing, please?

https://gist.github.com/atris/773e1074edd1d50786ff35e300495984

**Comment by atris** [comment]
@sijie @eolivelli Updated the PR, please see and comment

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/JournalChannel.java [hunk 1]
```diff
@@ -160,7 +176,13 @@ private JournalChannel(File journalDirectory, long logId,
                         + " suddenly appeared, is another bookie process running?");
             }
             randomAccessFile = new RandomAccessFile(fn, "rw");
-            fc = randomAccessFile.getChannel();
+
+            if (dummyFC == null) {
+                fc = randomAccessFile.getChannel();
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/JournalChannel.java]: What about using Powermock in order to hack the result of getChannel ?

I feel this way of testing is quite invasive

**Comment by atris** [comment]
The issue there would be that the RandomFileAccess instance used to create
the FileChannel is also created in the same context so we will probably
have to powermock up the call stack.

I felt this way is a longer term way to test different cases of FileChannel
implementations especially since FileChannel is an external class not owned
by our codebase.

Thoughts?
-- 
Regards,

Atri
*l'apprenant*

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/JournalChannel.java (hunk 1 above)
Comment (eolivelli) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/JournalChannel.java]: You can add a static method like:

FileChannel openFileChannel (RandomAccessFile)

Then you can override it with PowerMock

Comment (atris) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/JournalChannel.java]: @eolivelli Fixed, thanks

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java [hunk 2]
```diff
@@ -794,4 +801,50 @@ private void testPartialFileInfoPostV3Journal(boolean truncateMasterKey)
...
+        }
+    }
+
+    private class DummyJournalScan implements Journal.JournalScanner {
+        boolean printJournalVersion = false;
+
+        @Override
+        public void process(int journalVersion, long offset, ByteBuffer entry) throws IOException {
+            if (!printJournalVersion) {
+                System.out.println("Journal Version : " + journalVersion);
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: Use Logger please

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java [hunk 3]
```diff
@@ -794,4 +801,50 @@ private void testPartialFileInfoPostV3Journal(boolean truncateMasterKey)
...
+                .thenThrow(new IOException());
+
+        PowerMockito.mockStatic(JournalChannel.class);
+        PowerMockito.when(JournalChannel.openFileChannel(Mockito.any(RandomAccessFile.class))).thenReturn(fileChannel);
+
+        Bookie b = new Bookie(conf);
+        for (Journal journal : b.journals) {
+            List<Long> journalIds = journal.listJournalIds(journal.getJournalDirectory(), null);
+
+            assert journalIds.size() == 1;
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: Use assertEquals

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/JournalChannel.java [hunk 4]
```diff
@@ -290,4 +291,12 @@ public void forceWrite(boolean forceMetadata) throws IOException {
             this.lastDropPosition = newDropPos;
         }
     }
+
+    public static FileChannel openFileChannel(RandomAccessFile randomAccessFile) {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/JournalChannel.java]: Add VisibleForTesting comment

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java [hunk 5]
```diff
@@ -794,4 +801,50 @@ private void testPartialFileInfoPostV3Journal(boolean truncateMasterKey)
             // correct behaviour
         }
     }
+
+    @Test(expected = IOException.class)
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: Is is too ambiguous.
We ate not sure that only the last line threw this exception.

Please move the assertion exactly around the expected invocation

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java [hunk 6]
```diff
@@ -794,4 +801,50 @@ private void testPartialFileInfoPostV3Journal(boolean truncateMasterKey)
...
+        Journal.JournalScanner journalScanner = new DummyJournalScan();
+        FileChannel fileChannel = PowerMockito.mock(FileChannel.class);
+
+        PowerMockito.when(fileChannel.position(Mockito.anyLong()))
+                .thenThrow(new IOException());
+
+        PowerMockito.mockStatic(JournalChannel.class);
+        PowerMockito.when(JournalChannel.openFileChannel(Mockito.any(RandomAccessFile.class))).thenReturn(fileChannel);
+
+        Bookie b = new Bookie(conf);
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: Shall we stop the Bookie in a finally block?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java [hunk 7]
```diff
@@ -794,4 +801,50 @@ private void testPartialFileInfoPostV3Journal(boolean truncateMasterKey)
...
+            List<Long> journalIds = journal.listJournalIds(journal.getJournalDirectory(), null);
+
+            assert journalIds.size() == 1;
+
+            journal.scanJournal(journalIds.get(0), Long.MAX_VALUE, journalScanner);
+        }
+    }
+
+    private class DummyJournalScan implements Journal.JournalScanner {
+        boolean printJournalVersion = false;
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: This flag looks useless

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java (hunk 5 above)
Comment (atris) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: This annotation works at a class level -- not sure how I can make it to the specific line?

**Comment by atris** [comment]
@eolivelli Updated, please see and let me know your thoughts

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java (hunk 5 above)
Comment (eolivelli) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: It is not at **class** level, but it works around this **method**

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java [hunk 8]
```diff
@@ -794,4 +801,53 @@ private void testPartialFileInfoPostV3Journal(boolean truncateMasterKey)
...
+                .thenThrow(new IOException());
+
+        PowerMockito.mockStatic(JournalChannel.class);
+        PowerMockito.when(JournalChannel.openFileChannel(Mockito.any(RandomAccessFile.class))).thenReturn(fileChannel);
+
+        Bookie b = null;
+
+        try {
+            b = new Bookie(conf);
+        } finally {
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: you should shut down the bookie after the " for (Journal journal : b.journals) {" loop

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java [hunk 9]
```diff
@@ -794,4 +801,53 @@ private void testPartialFileInfoPostV3Journal(boolean truncateMasterKey)
...
+            b.shutdown();
+        }
+
+        for (Journal journal : b.journals) {
+            List<Long> journalIds = journal.listJournalIds(journal.getJournalDirectory(), null);
+
+            assertEquals(journalIds.size(), 1);
+
+            journal.scanJournal(journalIds.get(0), Long.MAX_VALUE, journalScanner);
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: try {
    journal.scanJournal(journalIds.get(0), Long.MAX_VALUE, journalScanner) ;
   fail("should not be able to scan the journal")
}  catch (IOException err) {
   // expected
   // maybe you can perform some assertion about the err.getMessage() result, but maybe it depends on java version or locale
}

Comment (atris) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: I tried that but it seems too brittle a check for depending on the error message.

**Comment by atris** [comment]
@eolivelli Updated, please see and let me know your thoughts

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java [hunk 10]
```diff
@@ -794,4 +801,63 @@ private void testPartialFileInfoPostV3Journal(boolean truncateMasterKey)
...
+
+        PowerMockito.mockStatic(JournalChannel.class);
+        PowerMockito.when(JournalChannel.openFileChannel(Mockito.any(RandomAccessFile.class))).thenReturn(fileChannel);
+
+        Bookie b = null;
+
+        try {
+            b = new Bookie(conf);
+        } finally {
+            b.shutdown();
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: Why are you shutting down the bookie twice?

Comment (atris) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/BookieJournalTest.java]: Oops, sorry, slipped through the cracks. Updated

**Comment by atris** [comment]
@eolivelli Updated,  please see. On another note, I see that TestEntryLog has just started failing again (the PR checks passed on my last commit 2 hours ago). Did something change?

**Comment by eolivelli** [comment]
LGTM

**Comment by eolivelli** [comment]
@sijie do you have comments?
@merlimat you may be interested in this fix

**Comment by eolivelli** [comment]
@atris please rebase to latest master, we have fixed a problem with git checkout

**Comment by eolivelli** [comment]
@atris please rebase to latest master, we have updated the github action "checkout"

**Comment by atris** [comment]
@eolivelli DOne

**Comment by eolivelli** [comment]
Hopefully we can commit this patch

Pinging @sijie @ivankelly @jiazhai

**Comment by atris** [comment]
Can we merge now?

**Comment by karanmehta93** [comment]
Thank you @atris for fixing this.

**Comment by eolivelli** [comment]
Merged to master branch, thank you @atris 
sorry that it took so much time.

I hope that our CI will be in better shape soon

**Pull request merged**